### ISSUE TITLE
`NeighborLoader` fails to sample edges in a one-hop heterogeneous graph

### ISSUE THREAD

**Issue by EdisonLeeeee** [pyg-team/pytorch_geometric#7280]
### 🐛 Describe the bug

Motivated by #7270, `NeighborLoader` seems doesn't work for a simple one-hop heterogeneous graph. Here is the code to reproduce:
```python
import torch
from torch_geometric.data import HeteroData
from torch_geometric.loader import NeighborLoader


data = HeteroData()
data['author'].x = torch.ones(2, 10)
data['paper'].x = torch.ones(6, 10)
data['author', 'paper'].edge_index = torch.tensor([[0, 0, 0, 1, 1, 1], [0, 1, 2, 3, 4, 5]])

loader = NeighborLoader(data, num_neighbors=[-1], input_nodes='author')

>>> next(iter(loader))['author', 'paper'].edge_index
tensor([], size=(2, 0), dtype=torch.int64)
```

### Environment

* PyG version: master
* PyTorch version: 2.0
* OS: macos
* Python version: 3.9
* CUDA/cuDNN version: N/A
* How you installed PyTorch and PyG (`conda`, `pip`, source): pip
* Any other relevant information (*e.g.*, version of `torch-scatter`):

**Comment by rusty1s** [comment]
Can you clarify the issue here? The `NeighborLoader` will sample from destination nodes to source nodes. As such,
if the `input_nodes` is `author`, there are no edges to sample since there does not exist a `(paper, author)` edge type. For `input_nodes = paper`, the output seems to be absolutely correct, as we are sampling the `(0, 0)` connection from paper to author.

**Comment by cctony1** [comment]
Oh. Thank you. 

So does this mean if we are interested in the neighborhood around the source 'author' nodes (e.g., for classification of those nodes), we need to make this graph undirected to sample each author's neighborhood?

In other words, do we always need edges pointing in toward the seed nodes to use the NeighborLoader?

**Comment by rusty1s** [comment]
Yes, that's correct. It doesn't necessarily need to be undirected (can be just reversed as well), but the default message passing flow in PyG is from source nodes to destination nodes (destination nodes will aggregate the information), so the `NeighborLoader` will respect this and sample accordingly.

**Comment by cctony1** [comment]
Thanks so much. I'm really grateful for how responsive the PyG team has been on my questions this week!

**Comment by cctony1** [comment]
Since we're going from Destination to Source...

when I specify num_neighbors=[10,20] for the NeighborLoader, does the 10 or 20 refer to the seed nodes' nearest neighbors?

I had assumed 10 was nearest neighbors to seed, and 20 was the neighbors' neighbors. But since I had the message passing direction reversed in my head, I want to verify this detail.

**Comment by rusty1s** [comment]
We start sampling from seed nodes. In the first hop, we sample 10 neighbors, and then sample 20 neighbors from each direct neighbor.

**Comment by cctony1** [comment]
Thanks.
